src/configureAWS.ts

```diff
@@ -20,51 +20,49 @@ import * as aws from 'aws-sdk';
 import {logger} from './logger';
 import {AWSRegion} from './aws-regions';
 
-async function loadSharedIniFile(profile?: string) {
+function getCredentialsProviderChain(profile?: string) {
+  const hasDotAWS = (awsUserDir && fs.existsSync(awsUserDir));
+  const providers = [];
+  if (profile) {
+    if (profile.startsWith('arn:')) {
+      throw new Error(
+        'The impossible happened. getCredentialsProviderChain was passed an arn formatted profile');
+      // NOTE: arn formatted profiles are handled elsewhere
+    }
+    if (!hasDotAWS) {
+      throw new Error('AWS profile provided but ~/.aws/{config,credentials} not found.');
+    }
+    providers.push(() => new aws.SharedIniFileCredentials({profile}));
+  } else {
+    providers.push(() => new aws.EnvironmentCredentials('AWS'));
+    if (hasDotAWS) {
+      providers.push(() => new aws.SharedIniFileCredentials({profile}));
+    }
+    providers.push(() => new aws.EC2MetadataCredentials());
+  // TODO ^ if (aws.ECSCredentials.prototype.isConfiguredForEcsCredentials()) {
+  //   return new aws.ECSCredentials();
+  // }
+  }
+  return new aws.CredentialProviderChain(providers);
+}
+
+async function resolveCredentials(profile?: string) {
   // note, profile might be undefined here and that's fine.
-  const credentials = new aws.SharedIniFileCredentials({profile});
-  //await credentials.refreshPromise();
-  aws.config.credentials = credentials;
+  if (profile && profile.startsWith('arn:')) {
+    const masterCreds = await getCredentialsProviderChain(undefined).resolvePromise();
```

Comment:
This should just use `aws.CredentialProviderChain.defaultProviders`, then you can drop the weird `profile.startsWith('arn:')` check in `getCredentialsProviderChain()`.